### PR TITLE
[XPU] Enable FLASH_ATTN on XPU

### PR DESCRIPTION
**Pull request by yma11**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
This PR is to add XPU path in `FLASH_ATTN` backend.

## Test Plan
```
 DIFFUSION_ATTENTION_BACKEND=FLASH_ATTN  python ./examples/offline_inference/text_to_image/text_to_image.py
 pytest -s -v tests/diffusion/attention/test_flash_attn.py
```
## Test Result
image
```
=== Case 1: Padding Equivalence Test ===
Output A shape: torch.Size([1, 64, 8, 64])
Output B shape: torch.Size([1, 84, 8, 64])
Output B unpadded shape: torch.Size([1, 64, 8, 64])
Max absolute difference: 0.000000
Mean absolute difference: 0.000000
✓ Case 1 PASSED: Padded and unpadded outputs are very close!
PASSEDGPU cleanup disabled

tests/diffusion/attention/test_flash_attn.py::test_fa_vs_sdpa
=== PRE-TEST GPU CLEANUP ===
GPU cleanup disabled
INFO 02-11 08:51:58 [scheduler.py:226] Chunked prefill is enabled with max_num_batched_tokens=2048.
--- Running test: test_fa_vs_sdpa

=== Case 2: FA vs SDPA Comparison ===
Batch size: 2
FA output shape: torch.Size([2, 84, 8, 64])
SDPA output shape: torch.Size([2, 84, 8, 64])
Max absolute difference (valid region): 0.007812
Mean absolute difference (valid region): 0.000000
✓ Case 2 PASSED: FA and SDPA outputs are very close!
PASSEDGPU cleanup disabled

```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/attention/backends/flash_attn.py [hunk 1]
```diff
@@ -143,3 +144,74 @@ def forward_npu(
...
+                query,
+                key,
+                value,
+                cu_seqlens_q=cu_seqlens,
+                cu_seqlens_k=cu_seqlens,
+                max_seqlen_q=max_seqlen,
+                max_seqlen_k=max_seqlen,
+                causal=self.causal,
+                softmax_scale=self.softmax_scale,
+            )
```
Comment (gcanlin) [inline vllm_omni/diffusion/attention/backends/flash_attn.py]: Can't we reuse `forward_cuda` in `forward_xpu`?

File: tests/diffusion/attention/test_flash_attn.py [hunk 2]
```diff
@@ -265,8 +270,8 @@ def test_fa_vs_sdpa():
     print("Running FlashAttention Padding Tests...")
     print("=" * 60)
 
-    # Try to run CUDA tests
-    if torch.cuda.is_available():
+    # Try to run tests
+    if torch.cuda.is_available() or torch.xpu.is_available():
```
Comment (gcanlin) [inline tests/diffusion/attention/test_flash_attn.py]: How about removing `torch.xxx.is_available`, instead, extract a global bool var like:
```
is_xxx = current_omni_platform.is_cuda_alike() or current_omni_platform.is_xpu()
```

Comment (yma11) [reply tests/diffusion/attention/test_flash_attn.py]: Sure. We will enable more XPU UTs so may be we can add such a check in utils eventually.

File: vllm_omni/diffusion/attention/backends/flash_attn.py (hunk 1 above)
Comment (yma11) [reply vllm_omni/diffusion/attention/backends/flash_attn.py]: For case `attention_mask` is `None`,  xpu needs some specific data preparation before call `flash_attn_func`, which is actually `flash_attn_varlen_func` and seems not make sense to add a branch for it in `forward_cuda`.

**Comment by yma11** [comment]
@gcanlin please take a look again. cc @xuechendi @zhenwei-intel

File: vllm_omni/diffusion/attention/backends/flash_attn.py (hunk 1 above)
Comment (gcanlin) [reply vllm_omni/diffusion/attention/backends/flash_attn.py]: Make sense. Thanks for the explanation!

**Review by gcanlin — APPROVED** [review]
LGTM

**Comment by yma11** [comment]
@hsliuustc0106 please take a look. Thanks.

File: tests/diffusion/attention/test_flash_attn.py [hunk 3]
```diff
@@ -15,6 +15,9 @@
 from vllm_omni.diffusion.attention.backends.abstract import AttentionMetadata
 from vllm_omni.diffusion.attention.backends.flash_attn import FlashAttentionImpl
 from vllm_omni.diffusion.attention.backends.sdpa import SDPAImpl
+from vllm_omni.platforms import current_omni_platform
+
+is_gpu = current_omni_platform.is_cuda_alike() or current_omni_platform.is_xpu()
```
Comment (hsliuustc0106) [inline tests/diffusion/attention/test_flash_attn.py:20]: current_omni_platform.is_cuda_alike() includes cuda and rocm?

File: tests/diffusion/attention/test_flash_attn.py [hunk 4]
```diff
@@ -265,8 +268,8 @@ def test_fa_vs_sdpa():
     print("Running FlashAttention Padding Tests...")
     print("=" * 60)
 
-    # Try to run CUDA tests
-    if torch.cuda.is_available():
+    # Try to run tests
+    if torch.cuda.is_available() or torch.xpu.is_available():
```
Comment (hsliuustc0106) [inline tests/diffusion/attention/test_flash_attn.py]: should this be changed to is_gpu as well?

File: tests/diffusion/attention/test_flash_attn.py (hunk 3 above)
Comment (yma11) [reply tests/diffusion/attention/test_flash_attn.py:20]: yes.
```
    def is_cuda_alike(self) -> bool:
        """Stateless version of [torch.cuda.is_available][]."""
        return self._enum in (PlatformEnum.CUDA, PlatformEnum.ROCM)
```

File: tests/diffusion/attention/test_flash_attn.py (hunk 4 above)
Comment (yma11) [reply tests/diffusion/attention/test_flash_attn.py]: This check is to assure GPU cards available, so I think not to change it is also fine?

image

**Comment by hsliuustc0106** [comment]
amd ci failed, can you check why?

**Comment by yma11** [comment]
> amd ci failed, can you check why?

Seems CI agent issue or network issue?
image

**Comment by yma11** [comment]
@hsliuustc0106 CI becomes green. Please take a look again.

**Review by lishunyang12 — COMMENTED** [review]
The padded branch in `forward_xpu` is a copy of `forward_cuda` -- worth extracting the shared logic instead of duplicating it.

File: vllm_omni/diffusion/attention/backends/utils/fa.py [hunk 5]
```diff
@@ -33,6 +33,9 @@
 elif current_omni_platform.is_xpu():
     try:
         from vllm.v1.attention.backends.fa_utils import flash_attn_varlen_func  # noqa: F401
+
+        # XPU uses unified flash_attn API, below assignment to ensure HAS_FLASH_ATTN true
```
Comment (lishunyang12) [inline vllm_omni/diffusion/attention/backends/utils/fa.py]: Setting `flash_attn_func = flash_attn_varlen_func` just to satisfy `HAS_FLASH_ATTN = flash_attn_func is not None` is pretty roundabout. `forward_xpu` never actually calls `flash_attn_func` -- it only uses `flash_attn_varlen_func`. Would it be cleaner to change the sentinel to `HAS_FLASH_ATTN = flash_attn_func is not None or flash_attn_varlen_func is not None` instead of this aliasing trick?

File: vllm_omni/diffusion/attention/backends/flash_attn.py [hunk 6]
```diff
@@ -143,3 +143,75 @@ def forward_npu(
...
+                    "causal": self.causal,
+                    "softmax_scale": self.softmax_scale,
+                },
+            )
+            if isinstance(out_unpad, tuple):
+                out_unpad = out_unpad[0]
+
+            out = _pad_input(out_unpad, indices_q, query.size(0), query_length)
+
+        else:
```
Comment (lishunyang12) [inline vllm_omni/diffusion/attention/backends/flash_attn.py]: The entire masked branch (lines 175-196) is identical to `forward_cuda`. If this needs to stay as a separate method (because the unmasked path differs), could you at least extract the shared padded-attention logic into a helper? Two copies of the same varlen dispatch will inevitably drift.

File: vllm_omni/diffusion/attention/backends/flash_attn.py [hunk 7]
```diff
@@ -143,3 +143,75 @@ def forward_npu(
...
+            from einops import rearrange
+
+            batch_size, q_len = query.size()[:2]
+            cu_seqlens = torch.arange(0, (batch_size + 1) * q_len, step=q_len, dtype=torch.int32, device=query.device)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/attention/backends/flash_attn.py]: `einops` is imported here at runtime but does not appear in the project requirements. Is it already guaranteed to be installed in the XPU environment? If so, it would be good to add it to the relevant requirements file. If not, the rearrange call is just `x.reshape(-1, *x.shape[2:])` and you can avoid the extra dependency entirely.

File: vllm_omni/diffusion/attention/backends/flash_attn.py [hunk 8]
```diff
@@ -143,3 +143,75 @@ def forward_npu(
...
+        value: torch.Tensor,
+        attn_metadata: AttentionMetadata = None,
+    ) -> torch.Tensor:
+        """XPU flash attention implementation."""
+        from vllm_omni.diffusion.attention.backends.utils.fa import (
+            HAS_FLASH_ATTN,
+            _pad_input,
+            _unpad_input,
+            _upad_input,
+            flash_attn_varlen_func,
```
Comment (lishunyang12) [inline vllm_omni/diffusion/attention/backends/flash_attn.py]: `forward_xpu` imports `flash_attn_func` from the utils but never uses it. Leftover from copying `forward_cuda`?

File: tests/diffusion/attention/test_flash_attn.py [hunk 9]
```diff
@@ -265,8 +268,8 @@ def test_fa_vs_sdpa():
     print("Running FlashAttention Padding Tests...")
     print("=" * 60)
 
-    # Try to run CUDA tests
-    if torch.cuda.is_available():
+    # Try to run tests
```
Comment (lishunyang12) [inline tests/diffusion/attention/test_flash_attn.py:271]: On a standard CUDA-only PyTorch build, `torch.xpu` does not exist as an attribute so this line would `AttributeError` before it ever returns `False`. The skipif decorators above already handle this via the `is_gpu` variable -- should this `__main__` block use `is_gpu` as well for consistency?

**Comment by hsliuustc0106** [comment]
please fix it asap, we expect it to be supported in next stable release

File: tests/diffusion/attention/test_flash_attn.py (hunk 9 above)
Comment (yma11) [reply tests/diffusion/attention/test_flash_attn.py:271]: This is a good point. Using `is_gpu` makes more sense.

File: vllm_omni/diffusion/attention/backends/flash_attn.py (hunk 7 above)
Comment (yma11) [reply vllm_omni/diffusion/attention/backends/flash_attn.py]: removed `einops` dependency.

**Comment by yma11** [comment]
@lishunyang12 made a refactor, can you help take a review again?

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer
